appcues/src/main/java/com/appcues/ui/AppcuesViewModel.kt

```diff
@@ -43,8 +47,32 @@ internal class AppcuesViewModel(
             val position: Int,
             val flatStepIndex: Int,
             val isPreview: Boolean,
-            val presentationComplete: ((ResultOf<Unit, com.appcues.statemachine.Error>) -> Unit),
-        ) : UIState()
+            private val completionHandler: ((ResultOf<Unit, com.appcues.statemachine.Error>) -> Unit),
+        ) : UIState() {
+            // used to track presentation state, to avoid calling the completion multiple times
+            private var presented: Boolean = false
+            private val mutex = Mutex()
+
+            // returns true if the completion handler was invoked, false if not (already presented)
+            suspend fun presentationComplete(result: ResultOf<Unit, com.appcues.statemachine.Error>): Boolean = mutex.withLock {
```

Comment:
note: we now wrap the usage of the completion handler, and use a `Mutex` for thread safety - as we need to ensure this callback only goes back to the state machine once. There is an edge case in the `onActivityChanged` usage where it _could_ be called twice - deep linking away once an experience has already rendered.